synapse/lib/fifo.py

```diff
@@ -0,0 +1,389 @@
+        self.wind = Window(self, path, xmit=xmit)
+        self.onfini(self.wind.fini)
+
+        task = s_glob.sched.loop(2, self.flush)
+        self.onfini(task.fini)
+
+        self._cullAtomBase(self.wind.bseq)
+
+    def ack(self, seqn):
+        '''
+        Acknowledge (and cull) an item in the sequence.
+        '''
+        retn = self.wind.ack(seqn)
+
+        if retn:
+            self._cullAtomBase(self.wind.bseq)
+
+        return retn
+
+    def _cullAtomBase(self, nseq):
+
+        with self.lock:
+
+            while self.seqs[0] < nseq:
+                fseq = self.seqs.pop(0)
+                path = self._getSeqPath(fseq)
+                os.unlink(path)
+
+    def resync(self, xmit=None):
+        '''
+        Re-synchronize this Fifo by sending all window entries.
+        (optionally specify a new xmit callback)
+
+        Args:
+            xmit (func): The fifo xmit() callback.
+
+        Example:
+
+            def xmit(qent):
+                seqn, nseq, item = qent
+                dostuff()
+
+            fifo.resync(xmit=xmit)
+        '''
+        return self.wind.resync(xmit=xmit)
+
+    def _findFifoAtom(self, nseq):
+        # if they specify an un-aligned sequence, find the prev
+        if nseq not in self.seqs:
+            nseq = [s for s in self.seqs if s <= nseq][-1]
+
+        return nseq, self.atoms.gen(nseq)
+
+    def _openFifoAtom(self, nseq):
+        '''
+        Open and return a new Fifo AtomFile by next sequence.
+        '''
+        if nseq not in self.seqs:
+            return None
+
+        return self.atoms.gen(nseq)
+
+    def _brefAtomCtor(self, nseq):
+        path = self._getSeqPath(nseq)
+        return s_atomfile.openAtomFile(path, memok=False)
+
+    def flush(self):
+        '''
+        Flush any file buffers associated with this Fifo.
+        '''
+        self.wind.flush()
+        [wind.flush() for wind in self.winds]
+
+    def put(self, item):
+        '''
+        Put a new item into the Fifo.
+
+        Args:
+            item (obj): The object to serialize into the Fifo.
+        '''
+        byts = s_msgpack.en(item)
+
+        with self.lock:
+
+            seqn = self.nseq
+
+            self.atom.writeoff(self.atom.size, byts)
+            self.nseq += len(byts)
+
+            if self.atom.size >= self.maxsize:
+                self.atom.fini()
+                self.seqs.append(self.nseq)
+                self.atom = self.atoms.gen(self.nseq)
+
+            qent = (seqn, self.nseq, item)
+
+            self.wind._may_put(qent)
+
+            # see if any of our readers are caught up...
+            [wind._may_put(qent) for wind in self.winds.vals()]
+
+    def _getPathJoin(self, *names):
+        dirn = self.getConfOpt('fifo:dir')
+        return s_common.genpath(dirn, *names)
+
+    @staticmethod
+    @s_config.confdef(name='fifo')
+    def _getFifoConf():
+        return (
+            ('fifo:dir', {'type': 'str', 'req': 1, 'doc': 'Path to the FIFO directory'}),
```

Comment:
We have a few places where we're pulling this config value. Do we want to set this as a object local value as well, so we can just refer to a local attribute?

Reply:
Actually, this one isn't perf sensitive ( since it's ultimately leading to someone doing a file operation ) so i left it off.